android/KMEA/app/src/main/java/com/tavultesoft/kmea/packages/LexicalModelPackageProcessor.java

```diff
@@ -0,0 +1,111 @@
+
+import java.io.File;
+import java.io.FileFilter;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * An extension of PackageProcessor
+ * A class of static methods for use in handling Keyman's .kmp file format, as it relates to the
+ * KMEA Engine. This is primarily for installing lexical model packages.
+ */
+public class LexicalModelPackageProcessor extends PackageProcessor {
+  private static final String TAG = "LMPackageProcessor";
+
+  public LexicalModelPackageProcessor(File resourceRoot) {
+    super(resourceRoot);
+  }
+
+  public File constructPath(File path, boolean temp) {
+    String kmpBaseName = getPackageID(path);
+    // Feel free to change this as desired - simply ensure it is unique enough to never be used as
+    // a legitimate package name.
+    String kmpFolderName = temp ? "." + kmpBaseName + ".temp" : kmpBaseName;
+
+    return new File(resourceRoot, KMManager.KMDefault_LexicalModelPackages + File.separator + kmpFolderName + File.separator);
+  }
+
+  private boolean lexicalModelExists(final String packageId, final String modelId) {
+    if (resourceRoot != null) {
+      FileFilter lexicalModelFilter = new FileFilter() {
+        @Override
+        public boolean accept(File pathname) {
+          if (pathname.isFile() && pathname.getName().startsWith(modelId) && FileUtils.hasLexicalModelExtension(pathname.getName())) {
+            return true;
+          }
+          return false;
+        }
+      };
+
+      File kmpFile = new File(packageId + ".kmp");
+      if (!kmpFile.exists()) {
+        kmpFile = new File(packageId + ".model.kmp");
+      }
+      File packageDir = constructPath(kmpFile, false);
+      File[] files = packageDir.listFiles(lexicalModelFilter);
+      if (files != null && files.length > 0) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  public Map<String, String>[] processEntry(JSONObject jsonEntry, String packageId) throws JSONException {
+    JSONArray languages = jsonEntry.getJSONArray("languages");
+
+    String modelId = jsonEntry.getString("id");
+    if (lexicalModelExists(packageId, modelId)) {
+      HashMap<String, String>[] models = new HashMap[languages.length()];
+
+      for (int i = 0; i < languages.length(); i++) {
+        models[i] = new HashMap<>();
+        models[i].put(KMManager.KMKey_PackageID, packageId);
+        models[i].put(KMManager.KMKey_LexicalModelName, jsonEntry.getString("name"));
+        models[i].put(KMManager.KMKey_LexicalModelID, jsonEntry.getString("id"));
+        models[i].put(KMManager.KMKey_LexicalModelVersion, jsonEntry.getString("version"));
+        models[i].put(KMManager.KMKey_LanguageID, languages.getJSONObject(i).getString("id"));
+        models[i].put(KMManager.KMKey_LanguageName, languages.getJSONObject(i).getString("name"));
+
+        if (welcomeExists(packageId)) {
+          File kmpFile = new File(packageId + ".kmp");
+          if (!kmpFile.exists()) {
```

Comment:
This pattern bothers me -- we shouldn't need to be doing this.

Reply:
and zapped